### DISCUSSION TITLE
How to use resourceQuery loader options with turbopack?

### DISCUSSION THREAD

**Discussion by clement-escolano** [vercel/next.js#65360] [Help] [answered]
### Summary

We recently migrated a big project to the latest version of Next and wanted to try out turbopack. The project loads faster with it so it feels good on this front. However, there is still one issue that blocks us from migrating completely to turbopack. We have special configuration for `@svgr/webpack`: almost all of our svg images are icons so we use the `icon: true` option with `@svgr/webpack`. However, some of our SVG files are actual images and we need to use these files as standard images.

With webpack, this configuration is done with:
<details>
  <summary>Webpack configuration</summary>

  ```js
webpack: (config) => {
    // use default url loader when importing SVG with "*.svg?url"
    config.module.rules.push({
        test: /\.svg$/i,
        type: "asset",
        resourceQuery: /url/,
    });

    config.module.rules.push({
        test: /\.svg$/i,
        issuer: /\.tsx?$/,
        resourceQuery: { not: [/url/] }, // only if not *.svg?url
        use: [
            {
                loader: "@svgr/webpack",
                options: {
                    icon: true,
                    svgProps: { role: "img", "aria-label": "" },
                },
            },
        ],
    });

    return config;
},
```
</details>

I did not find a way of doing it with turbo. This is the configuration I have for my icons but I can't add the `?url` special case so my standard SVG images won't work:
<details>
  <summary>Turbo configuration</summary>

```js
turbo: {
    rules: {
        "*.svg": {
            loaders: [
                {
                    loader: "@svgr/webpack",
                    options: {
                        icon: true,
                        svgProps: { role: "img", "aria-label": "" },
                    },
                },
            ],
            as: "*.js",
        },
    },
},
```
</details>

Do you know if this is currently feasible with turbopack?

Thank you

**Comment by RafaelFerreiraTVD** [comment]
@clement-escolano have you managed to find a solution for your issue? I need to achieve the same and I'm kinda blocked as well.

**Reply by clement-escolano** [reply]
No unfortunately. I have not investigated much but I think this will require a change on turbopack's end so we are stuck waiting for this one

**Reply by RafaelFerreiraTVD** [reply]
Hey @clement-escolano 

Not sure if this is of your interest, but for our use case we only needed 2 different ways of handling svgs, but I believe this solution is possible to extend.

We created a custom loader, and our next.config.js file is similar to the following:

```javascript
experimental: {
    turbo: {
      rules: {
        '*.svg': {
          loaders: [{
            loader: require.resolve('./path-to-custom-loader.js'), 
            options: {
              // dimensions: false,
              // ....
            }
          }],
          as: '*.js',
        },
      },
    },
  },
```

Then the code for the custom loader depends on your needs, but you can export as many named exports as you want, and you can use those in the import statement after.

```javascript
const { transform } = require('@svgr/core');

module.exports = function (source) {
  const callback = this.async();
  const options = this.getOptions();

  transform(source, {
  ...options,
  }).then((resultCode) => {
    // you can now export something similar to this:
    let result = `
    const base64string = "base64";
    export default base64string;
    export const MyComponent = base64string;
    export const MyComponent2 = () => {
        return <div>Hello World</div>;
    }`;

    // you can also append the "previous" ways with the transform from core:
    result = `${result}${resultCode}`;
    callback(null, result);
  });
};
```

Basically, as long as it's a valid javascript file result, it will work.
And now instead using the resourceQuery feature of webpack you will need to use the correct named export. Probably its not optimal, but I guess this will have small impact on performance.

`source` is the file content that you can pass to `@svgr/core` transform function

**Reply by clement-escolano** [reply]
Thank you for your answer and sorry for the late reply. Unfortunately, this does not work as the custom loader does not have access to the import string used for the SVG file so there is no way to tell which loader should be used.

**Reply by RafaelFerreiraTVD** [reply]
You now have access to the resourceQuery param and it was fixed here: https://github.com/vercel/next.js/pull/69703

**Reply by RafaelFerreiraTVD** [reply]
You can then try to either return 1 or another

**Reply by alex-statsig** [reply]
Is there any good way to conditionally use the default svg loader with this? My goal is to recreate the [default svgr setup for nextjs](https://react-svgr.com/docs/next/#usage) which skips the svgr loader if the "?url" resourceQuery is specified, so that you can easily import an "svg" as a react component (ex. for icons) or an asset (ex. for larger "images"). I tried this custom loader:
```
const plugin = require('@svgr/webpack');
const nextImagePlugin = require('next/dist/build/webpack/loaders/next-image-loader');

module.exports = function (source) {
  if (this.resourceQuery === '?url') {
    return nextImagePlugin.default.call(this, source);
  }
  return plugin.call(this, source);
};
```

but it seems like "next-image-loader" isn't really the default svg loader used with turbopack (instead its a native rust loader afaict).

Noticed [similar discussion](https://github.com/vercel/turborepo/issues/4832#issuecomment-2489518179) about this gap elsewhere, but the workaround isn't ideal

**Reply by vitalets** [reply, marked as answer]
Next.js 16 shipped Turbopack [condition rules](https://nextjs.org/docs/app/api-reference/config/next-config-js/turbopack#advanced-webpack-loader-conditions), which allow loaders to be applied conditionally.

For example, you can use the `path` condition to apply SVGR only to the `icons` folder while keeping the default loader for all other SVGs:

```ts
turbopack: {
  rules: {
    "*.svg": {
      loaders: ["@svgr/webpack"],
      condition: {
        path: "icons/**"
      },
      as: "*.js",
    },
  },
},
```

The only challenge is typings: TypeScript is not aware of Turbopack rules, so an imported `.svg` file might be treated either as a React component or as a static image `{ src, width, height }`.

In my projects, I replaced SVGR with inlining icons as data URIs. It’s more performant and also solves the typing issue. Imported icons always have the same shape as external images:

```js
{
  src: "data:image/svg+xml,...",
  width: 32,
  height: 32
}
```

Here is a [detailed write-up](https://dev.to/vitalets/turbopack-a-better-way-to-inline-svg-in-nextjs-16-36em).

**Reply by emmgfx** [reply]
@vitalets but with this solution we can't modify the color icon with css, right?

**Reply by vitalets** [reply]
@emmgfx For mono-color icons, I found a way to do it with [CSS masking](https://dev.to/vitalets/turbopack-a-better-way-to-inline-svg-in-nextjs-16-36em#problem-2-customizing-icon-color). For more fine-grained access to SVG internals, yes: SVGR or manually inlining that specific icon to get full control over its nodes.

**Reply by clement-escolano** [reply]
Thank you @vitalets This is not perfect but is the closest solution we can get right now

**Comment by Umuts-Codes** [comment]
Hi,


Currently, Turbopack does not fully support all the loader-specific query options available in Webpack, including the `?url` resource query for `@svgr/webpack`. This means that differentiating between SVG icons (with `icon: true`) and standard SVG images during the build is not directly supported out of the box.


### Recommended approaches:

1. **Separate Import Paths**

   * Keep SVG icons and standard SVG images in different folders.
   * Configure Turbopack or the Next.js project to handle each folder differently. For example, you can import standard SVGs using a dedicated wrapper that treats them as URLs, while importing icons using `@svgr/webpack`.



2. **Custom Wrapper Components**

   * For standard images, create a small wrapper component that appends the `?url` logic or uses `next/image`.
   * For icons, continue using the SVGR transformation.



3. **Fallback to Webpack for Complex Cases**

   * If your project relies heavily on `?url` queries for SVGs, it may be necessary to temporarily continue using Webpack for these assets until Turbopack fully supports loader query options.



- Summary -

* Turbopack currently lacks the Webpack-style resource query flexibility.
* The most reliable workaround is to separate SVGs by type and handle them explicitly in imports or with wrapper components.
* Monitor Next.js and Turbopack releases for future updates, as loader query support may improve in upcoming versions.

This approach ensures a clear separation between icons and image SVGs while maintaining compatibility with Turbopack’s current capabilities.

**Comment by jwueller** [comment]
#87291 is my attempt of solving this in alignment with the recently-added loader conditions. This is what your `turbopack` rule could look like if the PR gets merged:

```ts
"*.svg": {
    condition: { not: { query: /url/ } },
    loaders: [
        {
            loader: "@svgr/webpack",
            options: {
                icon: true,
                svgProps: { role: "img", "aria-label": "" },
            },
        },
    ],
    as: "*.js",
},
```
